test/python/synthesis/test_multi_controlled_synthesis.py

```diff
@@ -0,0 +1,384 @@
+        )
+        transpiled_circuit = pm.run(synthesized_circuit)
+        cx_count = transpiled_circuit.count_ops()["cx"]
+        # The algorithm synth_mcx_noaux_v24 is based on the synthesis of MCPhase,
+        # which is defined using a sequence of MCRZ gates:
+        # MCPhase(n) is defined using one MCRZ(1), one MCRZ(2), ..., one MCRZ(n).
+        # The bound below follows using the bound of 16*(k+1)-40 for MCRZ(k) and summing
+        # the resulting arithmetic progression:
+        #   sum_{k=1}^n (16*(k+1)-40) = sum_{k=1}^n (16*k - 24) =
+        #     16*n*(n+1)/2 - 24*n = 8n^2 - 16*n.
+        self.assertLessEqual(cx_count, 8 * num_ctrl_qubits**2 - 16 * num_ctrl_qubits)
+
+    @combine(
+        num_ctrl_qubits=[5, 8, 10, 13, 15],
+        base_gate=[
+            XGate(),
+            YGate(),
+            ZGate(),
+            HGate(),
+            SGate(),
+            SdgGate(),
+            TGate(),
+            TdgGate(),
+            SXGate(),
+            SXdgGate(),
+            PhaseGate(0.345),
+        ],
+    )
+    def test_mcx_equiv_noaux_cx_count(self, num_ctrl_qubits: int, base_gate: Gate):
+        """Test bounds on the number of CX-gates when synthesizing multi-controlled gates
+        which are locally equivalent to MCX.
+        """
+        qc = QuantumCircuit(num_ctrl_qubits + 1)
+        qc.append(base_gate.control(num_ctrl_qubits), range(num_ctrl_qubits + 1))
+        pm = generate_preset_pass_manager(
+            optimization_level=0, basis_gates=["u", "cx"], seed_transpiler=12345
+        )
+        transpiled_circuit = pm.run(qc)
+        cx_count = transpiled_circuit.count_ops()["cx"]
+        # The bounds should be the same as for synth_mcx_noaux_v24
+        self.assertLessEqual(cx_count, 8 * num_ctrl_qubits**2 - 16 * num_ctrl_qubits)
+
+    @data(5, 8, 10, 13, 15)
+    def test_mcu_noaux_cx_count(self, num_ctrl_qubits: int):
+        """Test bounds on the number of CX-gates when synthesizing multi-controlled single-qubit
+        unitary gates.
+        """
+        base_gate = UGate(0.123, 0.456, 0.789)
+        qc = QuantumCircuit(num_ctrl_qubits + 1)
+        qc.append(base_gate.control(num_ctrl_qubits), range(num_ctrl_qubits + 1))
+        pm = generate_preset_pass_manager(
+            optimization_level=0, basis_gates=["u", "cx"], seed_transpiler=12345
+        )
+        transpiled_circuit = pm.run(qc)
+        cx_count = transpiled_circuit.count_ops()["cx"]
+        # The synthesis of MCX(n) uses two MCRZ(n), one MCRY(n), and one MCPhase(n-1).
+        # Thus the number of CX-gate should be upper-bounded by
+        # 3*(16 * (n + 1) - 40) + (8 * (n-1)^2 - 16 * (n-1))
+        self.assertLessEqual(cx_count, 8 * num_ctrl_qubits**2 + 16 * num_ctrl_qubits - 96)
+
+    @combine(
+        num_ctrl_qubits=[1, 2, 3, 4, 5, 6, 7, 8],
+        base_gate=[
+            XGate(),
+            YGate(),
+            ZGate(),
+            HGate(),
+            PhaseGate(0.123),
```

Comment:
There are no numeric dependencies in these algorithms, right? If that's so how about using parameterized gates here to show the generality?

Reply:
Exactly! No numeric dependencies as far as I know.